cpp/src/sort/sort_impl.cuh

```diff
@@ -18,20 +18,226 @@
 
 #include <cudf/column/column_factories.hpp>
 #include <cudf/detail/gather.hpp>
-#include <cudf/strings/detail/sorting.cuh>
 #include <cudf/table/row_operators.cuh>
 #include <cudf/table/table_device_view.cuh>
 #include <cudf/utilities/error.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
-#include <rmm/device_vector.hpp>
+#include <rmm/device_uvector.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <cub/cub.cuh>
+
 #include <thrust/sequence.h>
+#include <thrust/sort.h>
 
 namespace cudf {
 namespace detail {
-// Create permuted row indices that would materialize sorted order
+
+/**
+ * @brief Type-dispatched functor for sorting a single column.
+ */
+template <bool stable = false>
+struct single_column_sort_fn {
+  /**
+   * @brief Comparator functor needed for stable sort.
+   */
+  template <typename T>
+  struct comparator {
+    __device__ bool operator()(size_type lhs, size_type rhs)
+    {
+      if (has_nulls) {
+        bool lhs_null{d_column.is_null(lhs)};
+        bool rhs_null{d_column.is_null(rhs)};
+        if (lhs_null || rhs_null) {
+          if (!ascending) thrust::swap(lhs_null, rhs_null);
+          return (null_precedence == cudf::null_order::BEFORE ? !rhs_null : !lhs_null);
+        }
+      }
+      auto const lh_elem = d_column.element<T>(lhs);
+      auto const rh_elem = d_column.element<T>(rhs);
+      return ascending ? lh_elem < rh_elem : rh_elem < lh_elem;
+    }
+    column_device_view const d_column;
+    bool has_nulls;
+    bool ascending;
+    null_order null_precedence{};
+  };
+
+  /**
+   * @brief Sorts numeric columns using the CUB Radix Sort.
+   *
+   * @param input Column to sort
+   * @param indices Output sorted indices
+   * @param ascending True if sort order is ascending
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   */
+  template <typename T, typename std::enable_if_t<cudf::is_numeric<T>()>* = nullptr>
+  void fast_numeric_sort(column_view const& input,
+                         mutable_column_view& indices,
+                         bool ascending,
+                         rmm::cuda_stream_view stream)
+  {
+    // A non-stable sort on a numeric column with no nulls can use the radix sort but
+    // requires making a copy of the input data and the indices.
```

Comment:
Would it be too complex to also add a fast-path directly to `cudf::sort` for these cases to avoid an extra copy?

Reply:
Added a fast-path for `cudf::sort` on a single column which requires no comparator. So this will be a fast sort without the memory overhead.